### ISSUE TITLE
question: rxExecuteBlocking() taking Promise

### ISSUE THREAD

**Issue by lwlee2608** [vert-x3/vertx-rx#213]
Given a scenario, I want to make some legacy synchronous API calls. In this example, I will simply call Thread.sleep() to represent that.

```
public String legacySyncApi() throws Exception{
    Thread.sleep(3000);
    return "ExampleResult";
}
```

In order not to block the event-loop, I will need to use rxExecuteBlocking api provided by vertx

```
vertx.rxExecuteBlocking( future -> {
    try {
        future.complete(legacySyncApi());
    } catch (Exception e) {
        future.fail(e);
    }
}) 
... chain to other rxjava operation
```

In our project, where we exclusively code with RxJava
Ideally, we would prefer

```
vertx.rxExecuteBlocking( () -> {
    return legacySyncApi();
})
```

or perhaps, chain it as method reference
```
vertx.rxExecuteBlocking(this::legacySyncApi);
```

In order to achieve this, we wrapped the rxExecuteBlocking with our own spin, replacing `Handler<Promise<T>>` with a more rxjava friendlier argument `Callable` (and also overload it with Runnable that so that it can return void)

```
public Maybe<Object> rxExecuteBlocking(Callable caller){
    return vertx.rxExecuteBlocking( objectFuture -> {
        try {
            objectFuture.complete(caller.call());
        } catch (Throwable e) {
            objectFuture.fail(e);
        }
    });
}
```

So my question is why is `rxExecuteBlocking` taking `Promise`, it seems a little bit unnecessary. I am guessing it is probably because of the original method `executeBlocking` return `void`?

**Comment by vietj** [comment]
likely the design of execute blocking in vertx-core should have used a `Callable<T>` instead of a `Handler<Promise<T>>`, I think `Handler<Promise<T>>` was chosen is because it can be code generated.

**Comment by vietj** [comment]
I think we could try overloading it with a Callable and see how it works with code generation of RX.

**Comment by lwlee2608** [comment]
Ah I see, I'll give it a try.

**Comment by lwlee2608** [comment]
@vietj You're right, vertx-codegen doesn't like Callable:
`Error: java:: type java.util.concurrent.Callable<T> is not legal for use for a parameter in code generation`
any plan of including it in the future?

**Comment by vietj** [comment]
@lwlee2608 unfortunately I think it would be a breaking change, I will see if we can do something in vertx-core

**Comment by lwlee2608** [comment]
That is unfortunate, let me know if I can help in any way.

**Comment by vietj** [comment]
@lwlee2608 it can be annotated with `@GenIgnore(GenIgnore.PERMITTED_TYPE)` to allow it

**Comment by lwlee2608** [comment]
Yeah, I've noticed that. But I wasn't sure if you're okay with it. Do you want me to create a PR for it?

**Comment by vietj** [comment]
I don't know yet.

I would like to check first if adding executeBlocking methods with Callable does not create an overloading conflict with the current Handler methods, if that does not, I think we can consider it.

**Comment by lwlee2608** [comment]
All right. Let me know. 

Besides `callable`, I would also like to overload it with `Runnable` since the rx APi return `Maybe<T>`

```
    <T> Maybe<T> rxExecuteBlocking(Runnable runnable) {
        return vertx.rxExecuteBlocking( promise-> {
            try {
                runnable.run();
                promise.complete();
            } catch (Throwable e) {
                promise.fail(e);
            }
        });
    }

    <T> Maybe<T> rxExecuteBlocking(Callable<? extends T> callable) {
        return vertx.rxExecuteBlocking( promise-> {
            try {
                promise.complete(callable.call());
            } catch (Throwable e) {
                promise.fail(e);
            }
        });
    }
```

**Comment by tsegismont** [comment]
This has been implemented in Vert.x 5
